File: src/main/java/io/github/tanguygab/conditionalactions/actions/ActionGroup.java
package io.github.tanguygab.conditionalactions.actions;

import io.github.tanguygab.conditionalactions.ConditionalActions;
import io.github.tanguygab.conditionalactions.actions.types.Action;
import io.github.tanguygab.conditionalactions.conditions.ConditionGroup;
import org.bukkit.OfflinePlayer;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class ActionGroup implements CAExecutable {

    private final ActionManager manager;
    private final List<CAExecutable> actions = new ArrayList<>();

    public ActionGroup(ActionManager manager, List<?> config) {
        this.manager = manager;
        config.forEach(action->{
            if (action instanceof Map<?,?> map) {
                ConditionGroup condition = ConditionalActions.getInstance().getConditionManager().getCondition((String) map.get("condition"));
                actions.add(new ActionCondition(condition,loadConditionList(map,"success"),loadConditionList(map,"deny")));
            } else add(actions,(String) action);
        });
    }

    private ActionGroup loadConditionList(Map<?,?> map, String name) {
        if (!map.containsKey(name)) return null;
        Object obj = map.get(name);
        List<?> list = obj instanceof List<?> l ? l : List.of(obj);
        return new ActionGroup(manager,list);
    }

    private void add(List<CAExecutable> list, String line) {
        if (line.equalsIgnoreCase("return")) {
            list.add(null);
            return;
        }
        Action ac = manager.find(line);
        if (ac != null) list.add(new ActionData(ac,line));
    }

    @Override
    public boolean execute(OfflinePlayer player, Map<String, String> replacements) {
        for (CAExecutable executable : actions) {
            if (executable == null || !executable.execute(player, replacements)) return false;
        }
        return true;
    }

}

File: src/main/kotlin/io/github/tanguygab/conditionalactions/actions/ActionGroup.kt
package io.github.tanguygab.conditionalactions.actions

import io.github.tanguygab.conditionalactions.ConditionalActions
import org.bukkit.OfflinePlayer
import kotlin.collections.get

class ActionGroup(
    private val manager: ActionManager,
    config: List<*>
) : CAExecutable {
    private val actions = mutableListOf<CAExecutable?>()

    init {
        config.forEach { action ->
            if (action is MutableMap<*, *>) {
                val condition = ConditionalActions.INSTANCE.conditionManager.getCondition(action["condition"] as String)
                actions.add(
                    ActionCondition(
                        condition = condition,
                        success = loadConditionList(action, "success"),
                        deny = loadConditionList(action, "deny")
                    )
                )
            } else add(actions, action as String)
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun loadConditionList(map: MutableMap<*, *>, name: String): ActionGroup? {
        val obj = map[name] ?: return null
        val list = obj as? List<Any> ?: listOf(obj)
        return ActionGroup(manager, list)
    }

    private fun add(list: MutableList<CAExecutable?>, line: String) {
        if (line.equals("return", ignoreCase = true)) {
            list.add(null)
            return
        }
        val ac = manager.find(line)
        if (ac != null) list.add(ActionData(ac, line))
    }

    override fun execute(player: OfflinePlayer?, replacements: Map<String, String>): Boolean {
        for (executable in actions) {
            if (executable == null || !executable.execute(player, replacements)) return false
        }
        return true
    }
}
